### PR TITLE
Support for COSE Receipts (MMRIVER)

### PR DESCRIPTION
**Pull request by robinbryce**
The receipt command produces COSE Receipts for any confirmed event or log entry.

These can be attached to scitt signed statements to make them transparent statements.

Depends on https://github.com/datatrails/go-datatrails-merklelog/pull/22

### CODE REVIEW

File: receipt.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+	"github.com/datatrails/go-datatrails-common/cbor"
+	"github.com/datatrails/go-datatrails-merklelog/massifs"
+	"github.com/urfave/cli/v2"
+)
+
+func NewReceiptCmd() *cli.Command {
+	return &cli.Command{
+		Name:  "receipt",
+		Usage: "Generate a COSE Receipt of inclusion for any merkllog entry",
```
Comment (honourfish) [inline receipt.go:19]:
```suggestion
		Usage: "Generate a COSE Receipt of inclusion for any merklelog entry",
```

File: cmd/veracity/main.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ func main() {
 	if envikwid == "1" || strings.ToLower(envikwid) == "true" {
 		ikwid = true
 	}
+	ikwid = true
```
Comment (henry739) [inline cmd/veracity/main.go:35]: Doesn't 'this force IKWID on all of the time?

File: receipt.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+				return err
+			}
+			tenantIdentity := cCtx.String("tenant")
+			if tenantIdentity == "" {
+				return fmt.Errorf("tenant identity is required")
+			}
+			log("verifying for tenant: %s", tenantIdentity)
+
+			var mmrIndex uint64
+			mmrIndex = cCtx.Uint64("mmrindex")
```
Comment (henry739) [inline receipt.go:69]: This appears to overwrite a named argument based on a positional one, if both are supplied. What's the reasoning behind that?

File: receipt.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+				receipt = []byte(hex.EncodeToString(cbor))
+			}
+
+			if cCtx.String("output") == "" {
+				n, err := os.Stdout.Write(receipt)
+				if err != nil {
+					return err
+				}
+				if n != len(receipt) {
+					return fmt.Errorf("failed to write all bytes to file")
```
Comment (henry739) [inline receipt.go:117]:
```suggestion
					return fmt.Errorf("failed to write all bytes to stdout")
```

File: replicatelogs.go [hunk 5]
```diff
@@ -503,11 +503,26 @@ func (v *VerifiedReplica) replicateVerifiedContext(
...
 			return fmt.Errorf("%w: %s, massif=%d", ErrRemoteLogInconsistentRootState, tenantIdentity, massifIndex)
 		}
 		return nil
 	}
 
 	return v.localReader.ReplaceVerifiedContext(remote, v.writeOpener)
 }
+
+func verifiedStateEqual(a *massifs.VerifiedContext, b *massifs.VerifiedContext) bool {
+	if len(a.Data) != len(b.Data) {
```
Comment (henry739) [inline replicatelogs.go:516]: Minor: Aren't we checking this again on line 506?

File: cmd/veracity/main.go (hunk 2 above)
Comment (robinbryce) [reply cmd/veracity/main.go:35]: ah yes, good spot. accidental dev hack

File: receipt.go (hunk 3 above)
Comment (robinbryce) [reply receipt.go:69]: convenient ux. if you prefer exclusively on or the other I can remote the positional ?

File: replicatelogs.go (hunk 5 above)
Comment (robinbryce) [reply replicatelogs.go:516]: Uh yeah. I can imporove that. The reason for the elevation is to avoid writing to disk when we know the remote is unchanged.

Comment (robinbryce) [reply replicatelogs.go:516]: I think I just figured the redundant check was a defence against future refactorings

**Review by henry739 — APPROVED** [review]
LGTM

**Review by henry739 — APPROVED** [review]
LGTM - Reviewed for code sanity / ux

**Pull request merged**